Publish blog post

greg0ire opened the pull request.

greg0ire committed:
Publish blog post

aartal commented on index.md:
```diff
@@ -0,0 +1,850 @@
+# Wiremock + testcontainers + Algolia + Go = ❤️
+
+When dealing with a SaaS like Algolia, testing can be awkward. Ideally, you
```
Effective intro that sets the stage by explaining the challenges and the need for a better solution 🙌

aartal commented on index.md:
```diff
@@ -0,0 +1,850 @@
+What happens in practice is that the SDK will use a polling mechanism to check
+if your task is published. This will result in a lot of similarly looking files.
+This is not very interesting to reproduce in your test, so I recommend simply
+deleting files representing a negative response. The files have names that are
```
The term "negative response" could be clarified with an example.

aartal commented on index.md:
```diff
@@ -0,0 +1,850 @@
+if your task is published. This will result in a lot of similarly looking files.
+This is not very interesting to reproduce in your test, so I recommend simply
+deleting files representing a negative response. The files have names that are
+a bit ugly, so I usually rename them for clarity.
```
Providing examples and renaming suggestions for file names would be helpful:

```suggestion
deleting files representing a negative response. The files have names that are
a bit ugly, so I usually rename them for clarity. 
For example, you might rename 
`1_indexes_test-index_task_226434943725-6e8689fa-9bbb-43fb-9d24-6824c02fc7d5.json`
to `index_test_task_published.json`.
```

aartal commented on index.md:
````diff
@@ -0,0 +1,850 @@
+/.env
+```
+
+### Writing the code under test
````
NIT: This could be clearer:

```suggestion
### Writing the code to be tested
```

greg0ire replied:
It was a reference to https://en.wikipedia.org/wiki/System_under_test, but you're right, let's reword this. The reference does not bring much value.

aartal commented on index.md:
````diff
@@ -0,0 +1,850 @@
+### Installing Go
+
+```console
+mise use go@1.24
````
In addition to installing / setting up the version of Go, this section could benefit from more detailed project setup instruction. For example, including commands like `go mod init module/path/indexer`.

aartal commented on index.md:
````diff
@@ -0,0 +1,850 @@
+```shell
+# .env
+ALGOLIA_APP_ID=your_app_id
+ALGOLIA_API_KEY=your_app_key
````
I'd advise clarifying that environment variable values need to be replaced with valid Algolia API ID/key sets.

aartal commented on index.md:
````diff
@@ -0,0 +1,850 @@
+functions.
+
+```go
+package indexer
````
Specifying each `.go` file path and filename would be helpful.

aartal commented on index.md:
````diff
@@ -0,0 +1,850 @@
+Algolia:
+
+```go
+package indexer
````
Same suggestion regarding filename and path.

aartal commented on index.md:
````diff
@@ -0,0 +1,850 @@
+
+We could interact with Wiremock by calling the REST API with the `net/http`
+package, but as it turns out, there is a dedicated SDK for that, and it supports
+recording since [this pull request I sent](https://github.com/wiremock/go-wiremock/pull/33).
+
+At the time of writing, the PR is merged but not released yet, so for now,
+let's use a commit hash:
+
+```console
+go get github.com/testcontainers/testcontainers-go@553afd3c51228cedff0fbbb744b6da94456dbfe1
````
The mentioned merged PR is in `github.com/wiremock/go-wiremock` but instructions to retrieve an untagged package version from a commit hash for `github.com/testcontainers/testcontainers-go` instead. Please clarify this discrepancy (it looks as indeed it should be go-wiremock).

greg0ire replied:
You're absolutely right, good catch!

aartal commented on index.md:
````diff
@@ -0,0 +1,850 @@
+will create json files.
+
+```go
+// indexer_test.go
````
It would be helpful to specify where the new pieces of updated code should go in each file. Consider including entire files with imports or providing a GitHub repo with commits for each stage of the demo.

greg0ire replied:
This PR contains all the files, but you're right, I will add a header linking to it.

greg0ire replied:
Currently in the process of doing the latter, and it helps me find issues in the blogpost. Thanks for this suggestion!

aartal commented on index.md:
````diff
@@ -0,0 +1,850 @@
+tc-wiremock.go:73: create container: container create: Error response from daemon: Invalid bind mount config: mount source "/builds/product-discovery/ms.indexer/internal/import/brandsuggestion/testdata" is forbidden by the allow list [/home /tmp] - update the bind mounts configuration and restart the agent to enable
+```
+
+It would seem that we cannot use a bind mount in the CI. Let us use our
+`record` constant to make the container options conditional:
````
It would be interesting to have an explanation as to why the bind mount configuration is an issue in the CI environment and how the solution addresses this problem.

greg0ire replied:
I must admit the cause is not entirely clear to me; searching for the error message on DuckDuckGo, Google or Github yields nothing, and I suspect it might come from Testcontainers could own configuration. Looking at [their docs](https://testcontainers.com/cloud/docs/#mounting-the-local-filesystem), I get even more confused because it states that mounting into countainers is not even implemented.

I will however elaborate on why the solution addresses the problem.

aartal commented on index.md:
```diff
@@ -0,0 +1,850 @@
+# Wiremock + testcontainers + Algolia + Go = ❤️
+
```
Consider adding a brief overview of what the article will cover and/or a table of contents.

greg0ire replied:
Addressed at the end of the first section.

aartal commented on index.md:
```diff
@@ -0,0 +1,850 @@
+	"github.com/algolia/algoliasearch-client-go/v4/algolia/search"
+)
+
+func TestIndexRecord(t *testing.T) {
```
Adding comments within the test code snippets to explain key parts of the code would make it easier for readers to understand what each part does 🙏

greg0ire replied:
I tried adding a few, if you see more places that lack comments, I'm interested.

aartal replied:
Added another comment with suggestions, resolving.

aartal reviewed (commented):
Altogether very nice and detailed step-by-step guide on how to set up and use Wiremock, testcontainers, and Algolia together. Easy to follow along 👏

greg0ire committed:
Use Go 1.24

greg0ire committed:
Create new Golang module

greg0ire committed:
Set up the environment

greg0ire committed:
Write the code to be tested

greg0ire committed:
Write the test with a real Algolia instance

greg0ire committed:
Increase number of retries

greg0ire commented on index.md:
```diff
@@ -0,0 +1,890 @@
+			search.NewEmptySearchMethodParams().SetRequests(
+				[]search.SearchQuery{
+					*search.SearchForHitsAsSearchQuery(
+						search.NewEmptySearchForHits().SetIndexName(indexName).SetQuery("time"),
```
Big blunder both I and Algolia need to fix.

```suggestion
						search.NewEmptySearchForHits().SetIndexName(indexName).SetQuery("test"),
```

greg0ire committed:
Add Wiremock in record mode

greg0ire committed:
Record test data

greg0ire committed:
Make recording conditional

greg0ire committed:
Drop "requiredScenarioState"

greg0ire committed:
Make it work in the CI

Avoid bind mounts in the CI. That should be fine because we only use
playback mode there.

greg0ire committed:
Extract generic code to test helpers

aartal commented on index.md:
```diff
@@ -0,0 +1,919 @@
+func TestIndexRecord(t *testing.T) {
+	debug.Enable() // helps with seeing the progress, since this is super long
+
+	appID := os.Getenv("ALGOLIA_APP_ID")
+	apiKey := os.Getenv("ALGOLIA_API_KEY")
+	indexName := "test-index"
+
+	client, err := search.NewClient(appID, apiKey)
+	if err != nil {
+		t.Fatalf("Failed to create client: %v", err)
+	}
+
+	// Create index indirectly by setting settings
+	_, err = client.SetSettings(client.NewApiSetSettingsRequest(
+		"test-index",
+		search.NewEmptyIndexSettings().SetSearchableAttributes([]string{"name"}),
+	))
+
+	if err != nil {
+		t.Fatalf("Failed to set settings: %v", err)
+	}
+
+	t.Cleanup(func() {
+		_, err := client.DeleteIndex(client.NewApiDeleteIndexRequest(indexName))
+		if err != nil {
+			t.Fatalf("Failed to delete index: %v", err)
+		}
+	})
+
+	indexRecord()
+
+	// Search for 'test'
+	searchResp, err := client.Search(
```
Regarding my previous suggestion to add more comments on the test code snippets: even if might feel redundant given the relatively short length of the snippet, I find that it adds readability to over-explain everything that is done, such as: 

```suggestion
        // Get Algolia credentials from environment variables
	appID := os.Getenv("ALGOLIA_APP_ID")
	apiKey := os.Getenv("ALGOLIA_API_KEY")
	indexName := "test-index"

        // Create a new Algolia client
	client, err := search.NewClient(appID, apiKey)
	if err != nil {
		t.Fatalf("Failed to create client: %v", err)
	}

	// Create index indirectly by setting settings
	_, err = client.SetSettings(client.NewApiSetSettingsRequest(
		"test-index",
		search.NewEmptyIndexSettings().SetSearchableAttributes([]string{"name"}),
	))
	if err != nil {
		t.Fatalf("Failed to set settings: %v", err)
	}

        // Ensure the test index is deleted after the test completes
	t.Cleanup(func() {
		_, err := client.DeleteIndex(client.NewApiDeleteIndexRequest(indexName))
		if err != nil {
			t.Fatalf("Failed to delete index: %v", err)
		}
	})

        // Call the function under test - this should index a record in Algolia
	indexRecord()

	// Verify the record was indexed by searching for it
        // This search should return records with "test" in their contents
	searchResp, err := client.Search(
```

Having said that, this is a personal preference. Comments can make the code look cluttered and increase scrolling, it can duplicate some of the post explanations, etc. I suggest adding them in the GitHub repo containing the demo code instead.

greg0ire replied:
Your comments are really great, and I think they will help readers. I  disseminated them through the blog post, adding them only on the first occurrence of code I duplicate, and adding more similar comments. I have also removed comments that are repeated when code is repeated, so that readers have an easier time noticing new code.

mmvaubrun commented on index.md:
```diff
@@ -0,0 +1,970 @@
+a bit ugly, so I usually rename them for clarity.
+For example, you might rename
+`1_indexes_test-index_task_226434943725-6e8689fa-9bbb-43fb-9d24-6824c02fc7d5.json`
+to `index_test_task_published.json`.
```
Maybe here having some Json content to illustrate the different status could make things a bit clearer

mmvaubrun commented on index.md:
```diff
@@ -0,0 +1,970 @@
+
+After dropping `"requiredScenarioState" : "scenario-1-1-indexes-test-index-task-226434943725-43",`
+from the mapping file about polling, the test passes again, only this time, it
+passes in under 2 seconds.
```
I think here we would benefit from a whole section about scenarios and how Wiremock handle it as it is a core thing for stateful/replay mechanism

greg0ire replied:
The blogpost is already a bit long and besides, I'm still very unfamiliar with Wiremock and its concepts like scenarios, so I won't venture in this territory for now.

mmvaubrun replied:
Yes but as the replay feature is based on the stateful behaviour I think it is worth mentioning quickly this powerful: 

> It is possible for request to mention a scenario they belong to, allowing to do things like "On first 2 calls respond A and on 3rd return B".  Based on that it is possible to build complete choreography of request/response

greg0ire replied:
Thanks, added!

mmvaubrun commented on index.md:
```diff
@@ -0,0 +1,970 @@
+
+When recording, we mount the volume, which is not an issue because we are not
+in the CI.
+Otherwise, we use the `WithMappingFile` function which relies on a copy operation.
```
An addition to testContainer Wiremock facade could be helpful here to explain why it is so easy to do, and that other things are also possible

greg0ire replied:
I'll add a paragraph about that, but I think I don't think I will mention the rest of the API exposed by this facade, as it is very small. There are not a lot of other helpful functions.

mmvaubrun replied:
Ok maybe a link to do then.
My though was to add some information that is a provided APIs an not some dark magic 😄

greg0ire replied:
👍 just added a link

greg0ire committed:
Add images

greg0ire committed:
Drop numerous notPublished steps
